### ISSUE TITLE
Support Extensions in custom protocols

### ISSUE THREAD

**Issue by RangerMauve** [electron/electron#26812]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

### Problem Description

I'm making heavy usage of custom protocol schemes for loading content from p2p protocols as websites in my [Agregore Web Browser](https://github.com/AgregoreWeb/agregore-browser).

I'm currently making heavy use of extensions via an old branch of the electron-extensions module, but I'd like to transition to either using the built in extension support, or something that makes use of it.

Sadly, it looks like the built in extensions are unable to run content scripts or anything else on pages served from custom protocol schemes.

### Proposed Solution

It looks like we need to add something which enables the [SCHEME_ALL](https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/url_pattern.h;l=52;bpv=1;bpt=1?q=valid_schemes&ss=chromium%2Fchromium%2Fsrc) flag for valid schemes in `<all_urls>` in extensions, or have some additional code for validating schemes.

Maybe we'd be able to detect the `<all_urls>` flag in the [URLPatternSet](https://github.com/electron/electron/search?q=URLPatternSet) of an extension and enable the flag there?

Sadly it feels like Chrome's extension/custom protocol support really isn't cut out for this use case is valid_schemes is a bitfield rather than a set. 😅 

### Alternatives Considered

Maybe we could add a custom [url_pattern_set](https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/url_pattern_set.h;bpv=1;bpt=1) which we can set on extensions which adds support for a set of custom protocols?

Maybe the [CreateUserScript](https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/manifest_handlers/content_scripts_handler.cc;l=106;bpv=1;bpt=1?q=CreateUserScript&ss=chromium%2Fchromium%2Fsrc) code could be modified to account for custom protocols?

It could also be that I misunderstood that `<all_urls>` isn't matching custom protocols, since it looks like it [returns true as soon as it sees it regardless of valid schemes](https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/url_pattern.cc;l=202;drc=d81c5852498699fe3cd812e78d31c77c28e29281;bpv=1;bpt=1) 

### Additional Information

I've created a minimal repo that reproduces this issue here: https://github.com/RangerMauve/electron-custom-protocols-extension-test

You can see that for the `https://` scheme, the page gets modified with an extra "hello world" at the end, but for the custom `example://` scheme, it doesn't. This is with the extension having `<all_urls>` enabled.

**Comment by RangerMauve** [comment]
Gonna look into this again. Would a PR and review be welcome for this functionality?

**Comment by Slapbox** [comment]
@MarshallOfSound is there any plan to address this? Is it even possible to do so without changes upstream in Chromium?

**Comment by RangerMauve** [comment]
From what I looked I'm pretty sure the url_pattern is what will need the most changes which will require chromium patches. I don't know if it's realistic to move upstream.

This might be something that's relevant to the folks at Igalia though since they're working on integrating IPFS protocol handlers into Chromium, I assume it'd be useful for extensions to affect pages served from the handlers.

**Comment by Slapbox** [comment]
Friendly bump.

I assume the Electron team has at least a little contact with the Chromium team? I wonder if you could raise this issue with them if it's blocked upstream? 

@ckerr can you comment on this? Given the locking down of Electron - developers still do need these tools.

It's a difficult time to write a performant Electron application in a timely fashion because a number of tools are no longer accessible to us.

**Comment by Slapbox** [comment]
Could we get any comment on whether this might ever happen? Developing a React app without being able to profile it in production is _very bad_, to use the technical term.

**Comment by yannick-was-taken** [comment]
This is a very important feature. Right now you have to decide between either a custom protocol or devtools and other extensions. Even a workaround akin to enabling all schemes for an extension from the main process would be of great help.

**Comment by Slapbox** [comment]
I agree that this really needs some attention from the maintainers. Us end-developers can't get in touch with the Chromium dev team and voice these concerns in the same way the maintainers of Electron can.

**Comment by RangerMauve** [comment]
FWIW I made some progress on this by reimplementing a bunch of the WebExtensions spec for my web browser Agregore. https://github.com/AgregoreWeb/electron-extended-WebExtensions

I don't have devtools extensions in there yet but I'm open to PRs.

**Comment by Slapbox** [comment]
Thanks for sharing @RangerMauve!

Am I understanding correctly that features not marked `(doesn't work on custom protocols yet)` should work with custom protocols? How is that achieved?

**Comment by RangerMauve** [comment]
Yeah, it works by implementing the extension APIs in JavaScript and overwriting parts of the built in extensions apis.

The bulk of the logic id [here](https://github.com/AgregoreWeb/electron-extended-WebExtensions/blob/main/index.js) in the main thread, then I inject a preload script that's defined [here](https://github.com/AgregoreWeb/electron-extended-WebExtensions/blob/main/preloads-raw.js)

**Comment by tmm1** [comment]
> I've created a minimal repo that reproduces this issue here: https://github.com/RangerMauve/electron-custom-protocols-extension-test

Thanks for this repro. I was able to make it work with this very hacky patch:

```diff
diff --git a/extensions/common/url_pattern.cc b/extensions/common/url_pattern.cc
index 1b233c1597e26..c5ce4e8511812 100644
--- a/extensions/common/url_pattern.cc
+++ b/extensions/common/url_pattern.cc
@@ -146,7 +146,7 @@ bool URLPattern::IsValidSchemeForExtensions(std::string_view scheme) {
       return true;
     }
   }
-  return false;
+  return true;
 }

 // static
@@ -458,7 +458,7 @@ bool URLPattern::MatchesURL(const GURL& test) const {
   // Ensure the scheme matches first, since <all_urls> may not match this URL if
   // the scheme is excluded.
   if (!MatchesScheme(test_url->scheme_piece())) {
-    return false;
+    //return false;
   }

   if (match_all_urls_) {
```

**Comment by nikwen** [comment]
Here we go! I have a patch to make DevTools extensions work with custom protocols.

So far, the code is somewhat hacky. I’ll need to clean it up, expose proper APIs, add tests, etc. But it’s working and the path forward is clear.

@tmm1's comment helped. 🙏

<img width="2658" height="1812" alt="Image" src="https://github.com/user-attachments/assets/964162db-5cc0-46c6-87da-285267bb419f" />

**Comment by nikwen** [comment]
Posted a PR: https://github.com/electron/electron/pull/49951
